File: 3444_minimum_increments_for_target_multiples_in_an_array/solution.go
// LeetCode 3444 - Minimum Increments for Target Multiples in an Array
// https://leetcode.com/problems/minimum-increments-for-target-multiples-in-an-array/

func minimumIncrements(nums []int, target []int) int {
	m := len(target)
	N := 1 << m
	const inf = int(1e18)
	dp := make([]int, N)
	for i := range dp {
		dp[i] = inf
	}
	dp[0] = 0
	for _, x := range nums {
		ndp := append([]int(nil), dp...)
		for mask := 0; mask < N; mask++ {
			// assign x (after increment) to cover subset of targets
			for sub := 1; sub < N; sub++ {
				lcm := 1
				ok := true
				for i := 0; i < m; i++ {
					if sub&(1<<i) != 0 {
						lcm = lcm3444(lcm, target[i])
						if lcm > 1e9 {
							ok = false
							break
						}
					}
				}
				if !ok {
					continue
				}
				// cost to make x multiple of lcm
				cost := (lcm - x%lcm) % lcm
				nmask := mask | sub
				if dp[mask]+cost < ndp[nmask] {
					ndp[nmask] = dp[mask] + cost
				}
			}
		}
		dp = ndp
	}
	return dp[N-1]
}

func gcd3444(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}
func lcm3444(a, b int) int {
	return a / gcd3444(a, b) * b
}

File: 3444_minimum_increments_for_target_multiples_in_an_array/solution.rb
# LeetCode 3444 - Minimum Increments for Target Multiples in an Array
# https://leetcode.com/problems/minimum-increments-for-target-multiples-in-an-array/

# @param {Integer[]} nums
# @param {Integer[]} target
# @return {Integer}
def minimum_increments(nums, target)
  m = target.length
  nmask = 1 << m
  inf = 10**18
  dp = Array.new(nmask, inf)
  dp[0] = 0
  nums.each do |x|
    ndp = dp.dup
    (0...nmask).each do |mask|
      (1...nmask).each do |sub|
        l = 1
        ok = true
        (0...m).each do |i|
          next if (sub & (1 << i)) == 0

          l = lcm_3444(l, target[i])
          if l > 1_000_000_000
            ok = false
            break
          end
        end
        next unless ok

        cost = (l - x % l) % l
        nm = mask | sub
        ndp[nm] = dp[mask] + cost if dp[mask] + cost < ndp[nm]
      end
    end
    dp = ndp
  end
  dp[nmask - 1]
end

def gcd_3444(a, b)
  while b != 0
    a, b = b, a % b
  end
  a
end

def lcm_3444(a, b)
  a / gcd_3444(a, b) * b
end
